core/src/main/java/org/infinispan/statetransfer/CommitManager.java

```diff
@@ -68,11 +69,41 @@ public final void stopTrack(Flag track) {
          }
          tracker.clear();
       } else {
-         for (Iterator<Map.Entry<Object, DiscardPolicy>> iterator = tracker.entrySet().iterator();
-              iterator.hasNext(); ) {
-            if (iterator.next().getValue().update(trackStateTransfer, trackXSiteStateTransfer)) {
-               iterator.remove();
-            }
+         for (Map.Entry<Integer, Map<Object, DiscardPolicy>> segment : tracker.entrySet()) {
+            clearSegmentKeys(segment.getKey(),
+                  policy -> policy.getValue().update(trackStateTransfer, trackXSiteStateTransfer));
+         }
+      }
+   }
+
+   /**
+    * Stop tracking the entries for the given segment if tracking is enabled.
+    *
+    * @param flag: flag to verify if tracking is enabled.
+    * @param segmentId: segment to stop tracking.
+    */
+   public final void stopTrackFor(Flag flag, int segmentId) {
+      if (isTracking(flag)) {
+         // We only remove entries that are not related to cross-site state transfer. Different sites may have
+         // different configurations, thus a single entry may have different segment mapping varying from site to site.
+         clearSegmentKeys(segmentId, policy -> policy.getValue().discardXSiteST);
```

Comment:
this does not sound right. You are removing the `DiscarPolicy` if `discardXSiteST` is `true` and you shouldn't.